src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,66 +34,80 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        internal static T EnsureInitialized<T>([NotNull] ref T? field, T value) where T : class
+            // Interlocked.CompareExchange ensures we don't overwrite a field that was set.
+            => Interlocked.CompareExchange(ref field, value, null) ?? value;
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+        internal static T EnsureInitializedDisposable<T>([NotNull] ref T? field, T value) where T : class, IDisposable
+        {
+            if (Interlocked.CompareExchange(ref field, value, null) != null)
+            {
+                value.Dispose();
+            }
+
+            return field;
+        }
+
+        public static TextReader In => Volatile.Read(ref s_in) ?? ConsolePal.EnsureInitializedIn(ref s_in);
 
         public static Encoding InputEncoding
         {
             get
             {
-                return EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
+                return Volatile.Read(ref s_inputEncoding) ?? EnsureInitializedInputEncoding();
+
+                static Encoding EnsureInitializedInputEncoding()
+                    => EnsureInitialized(ref s_inputEncoding, ConsolePal.InputEncoding);
             }
             set
             {
                 CheckNonNull(value, nameof(value));
-                lock (InternalSyncObject)
-                {
-                    // Set the terminal console encoding.
-                    ConsolePal.SetConsoleInputEncoding(value);
 
-                    Volatile.Write(ref s_inputEncoding, (Encoding)value.Clone());
+                // Set the terminal console encoding.
```

Comment:
If somebody called `InputEncoding` from multiple threads in parallel, we would always end up in consistent state before this change. We can end up in inconsistent state after this change (e.g. the PAL can be set to value from one thread, and `s_inputEncoding` to value from other thread.)

Reply:
Ah yes, I didn't consider concurrent setting. I'll change back to using the lock here.